aten/src/ATen/cudnn/Descriptors.cpp

```diff
@@ -125,7 +125,16 @@ void FilterDescriptor::set(const at::Tensor &t, int64_t pad) {
     size[i] = (int) 1;
   }
   dim = std::max(dim, pad);
-  set(getDataType(t), (int) dim, size);
+  cudnnTensorFormat_t filter_format = CUDNN_TENSOR_NCHW;
+  // Implementation of `suggest_memory_format()` is ambiguous in cases:
+  // a. Tensor with channel size == 1
+  // b. Tensor with both spatial size == 1
+  // It causes mismatch memory format for data & filter in convolution. Hence we
+  // check for contiguous here to fallback to NCHW in those cases.
+  if (!t.is_contiguous() && t.suggest_memory_format() == at::MemoryFormat::ChannelsLast) {
```

Comment:
No need to check `!t.is_contiguous()` as it is mutually exclusive with `t.suggest_memory_format() == at::MemoryFormat::ChannelsLast`

Reply:
This is the special case as the comment above suggests.

They are not ~mutually exclusive~ identical (I believe you were actually saying this) in the two cases listed above.

I agree that it would be a better ideal to not have it here and put it inside `TensorImpl` instead, we could have the tag mutually exclusive here: https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L1514-L1515

Reply:
Is this an oversight or I missed something else: https://github.com/pytorch/pytorch/blob/32efb431294d99a60899b0809c6363065608e556/c10/core/TensorImpl.h#L1524-L1546.
But even after patching that my fix is still not working. :/ For some reason, calling `contiguous(channels_last)` on a non-contiguous (NC11) tensor ended up with both flags to be True. Tracing this down the code path, I am stepping on `empty_like`, which I know you are working on in a different thread. I'll double check the issue I mentioned earlier after your #23899 .

Put a second thought on this, the flags are not necessarily exclusive. I fallback to your suggestion earlier to update fallback to ChannelsLast in `suggest_memory_format` instead (just like my code is doing right now)
But this means that we cannot represent NC11 kernel in NHWC flag. I'll open an issue to track this in case we want to revisit our design later.